File: dart/binary_trees/is_same_tree.dart
import 'tree_node.dart';

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   int val;
 *   TreeNode? left;
 *   TreeNode? right;
 *   TreeNode([this.val = 0, this.left, this.right]);
 * }
 */

class Solution {
  bool isSameTree(TreeNode? a, TreeNode? b) {
    // If both trees are empty, return true
    if (a == null && b == null) {
      return true;
    }

    // If both tree are not empty, check it corresponding value and recursion
    // check their left and right
    if (a != null && b != null) {
      return a.val == b.val &&
          isSameTree(a.left, b.left) &&
          isSameTree(a.right, b.right);
    }
    // If one of them is empty, return false
    return false;
  }
}

File: leetcode/binary_trees/go/is_same_tree.go
type TreeNode struct {
	Val int
	Left *TreeNode 
	Right *TreeNode 
}


func isSameTree(a TreeNode, b TreeNode) bool {
	if a == nil || b == nil {
		return a == b
	}
	return a.Val == b.Val && isSameTree(a.Left, b.Left) && isSameTree(a.Right, b.Right)
}